archive/100-theorems-list/81_sum_of_prime_reciprocals_diverges.lean

```diff
@@ -0,0 +1,281 @@
+      rw hf_def,
+      simpa } },
+    simpa },
+
+  calc card M₁ ≤ card (image f K)                    : card_le_of_subset h
+  ...          ≤ card K                              : card_image_le
+  ...          ≤ 2 ^ card (image nat.succ (range k)) : by simp only [K, card_powerset]
+  ...          ≤ 2 ^ card (range k)                  : pow_le_pow one_le_two card_image_le
+  ...          = 2 ^ k                               : by rw (card_range k),
+end
+
+lemma lemma4_card_M_le_2_pow_k_mul_sqrt_x {x k : ℕ} :
+  card {e ∈ range x | ∀ p : ℕ, (nat.prime p ∧ p ∣ (e + 1)) → p ≤ k } ≤ (2 ^ k) * nat.sqrt(x) :=
+begin
+  let M := {e ∈ range x | ∀ p : ℕ, (nat.prime p ∧ p ∣ (e + 1)) → p ≤ k },
+  let M₁ := {e ∈ range x | squarefree (e + 1) ∧ ∀ p : ℕ, (nat.prime p ∧ p ∣ (e + 1)) → p ≤ k},
+  let M₂ := {e ∈ range (nat.sqrt x) | ∀ p : ℕ, (nat.prime p ∧ p ∣ (e + 1)) → p ≤ k},
+  let K := finset.product M₁ M₂,
+  let f : ℕ × ℕ → ℕ := λ mn, (mn.2 + 1) ^ 2 * (mn.1 + 1) - 1,
+
+  have h1 : M ⊆ image f K,
+  { intros m hm,
+    simp only [M, sep_def, mem_filter, mem_range] at hm,
+    obtain ⟨hmx, hmp⟩ := hm,
+    have h1' : ∃ (a b : ℕ), (a ∈ M₁ ∧ b ∈ M₂) ∧ f (a, b) = m,
+    { obtain ⟨a, b, hab₁, hab₂⟩ := nat.sq_mul_squarefree_of_pos' (nat.zero_lt_succ m),
+
+      have h11 : a ∈ M₁,
+      { simp only [M₁, sep_def, mem_filter, mem_range],
+        have ham : (a + 1) ∣ (m + 1) := dvd.intro_left ((b + 1) ^ 2) hab₁,
+        refine ⟨(lt_of_le_of_lt _) hmx, hab₂, _⟩,
+        { exact nat.succ_le_succ_iff.mp ((nat.le_of_dvd (nat.zero_lt_succ m)) ham) },
+        { intro p,
+          specialize hmp p,
+          rintros ⟨hpp, hpa⟩,
+          exact hmp ⟨hpp, dvd.trans hpa ham⟩ } },
+
+      have h12 : b ∈ M₂,
+      { simp only [M₂, sep_def, mem_filter, mem_range],
+        have hbm₁ : (b + 1) ^ 2 ∣ (m + 1) := dvd.intro (a + 1) hab₁,
+        have hbm₂ : (b + 1) ∣ (m + 1) := nat.dvd_of_pow_dvd one_le_two hbm₁,
+        split,
+        { calc b < b + 1 : lt_add_one b
+          ...    ≤ nat.sqrt(m + 1) : by { rw [nat.le_sqrt, ← pow_two (b + 1)],
+                                          exact nat.le_of_dvd (nat.zero_lt_succ m) hbm₁ }
+          ...    ≤ nat.sqrt x : nat.sqrt_le_sqrt (nat.succ_le_iff.mpr hmx) },
+        { intro p,
+          specialize hmp p,
+          rintros ⟨hpp, hpb⟩,
+          exact hmp ⟨hpp, dvd.trans hpb hbm₂⟩ } },
+
+      have h13 : f (a, b) = m,
+      { have h13' : (b + 1) ^ 2 * (a + 1) - 1 = m, rw [hab₁, nat.succ_sub_one m],
+        simpa [f] },
+
+      exact ⟨a, b, ⟨h11, h12⟩, h13⟩ },
+    simpa },
+
+  have h2 : card M₁ ≤ 2 ^ k := lemma4_aux_card_M1_le_2_pow_k,
+  have h3 : card M₂ ≤ nat.sqrt x,
+  { rw ← card_range (nat.sqrt x), refine card_le_of_subset _, simp [M₂] },
+
+  calc card M ≤ card (image f K)   : card_le_of_subset h1
+  ...         ≤ card K             : card_image_le
+  ...         = card M₁ * card M₂  : card_product M₁ M₂
+  ...         ≤ 2 ^ k * nat.sqrt x : mul_le_mul' h2 h3,
```

Comment:
I'd like to suggest rearranging this part of the proof. I think things will be much easier to follow this way.
```suggestion
    simp only [M, M₁, M₂, mem_image, exists_prop, prod.exists, mem_product, sep_def, mem_filter,
              mem_range] at hm ⊢,
    have hm' := m.zero_lt_succ,
    obtain ⟨a, b, hab₁, hab₂⟩ := nat.sq_mul_squarefree_of_pos' hm',
    obtain ⟨ham, hbm⟩ := ⟨dvd.intro_left _ hab₁, dvd.intro _ hab₁⟩,
    refine ⟨a, b, ⟨⟨_, hab₂, λ p hp, _⟩, ⟨_, λ p hp, _⟩⟩, by simp_rw [f, hab₁, m.succ_sub_one]⟩,
    { exact (nat.succ_le_succ_iff.mp (nat.le_of_dvd hm' ham)).trans_lt hm.1 },
    { exact hm.2 p ⟨hp.1, dvd.trans hp.2 ham⟩ },
    { calc b < b + 1         : lt_add_one b
      ...    ≤ (m + 1).sqrt  : by simpa only [nat.le_sqrt, pow_two] using nat.le_of_dvd hm' hbm
      ...    ≤ x.sqrt        : nat.sqrt_le_sqrt (nat.succ_le_iff.mpr hm.1) },
    { exact hm.2 p ⟨hp.1, dvd.trans hp.2 (nat.dvd_of_pow_dvd one_le_two hbm)⟩ } },

  have h2 : card M₂ ≤ nat.sqrt x,
  { rw ← card_range (nat.sqrt x),
    apply card_le_of_subset,
    simp [M₂] },

  calc card M ≤ card (image f K)   : card_le_of_subset h1
  ...         ≤ card K             : card_image_le
  ...         = card M₁ * card M₂  : card_product M₁ M₂
  ...         ≤ 2 ^ k * x.sqrt     : mul_le_mul' lemma4_aux_card_M1_le_2_pow_k h2,
```